docs/source/hallway.rst

```diff
@@ -0,0 +1,363 @@
+===========================
+
+The ``graphenv`` library implements graph search as a reinforcement
+learning (RL) problem with parameteric action spaces, and is ready to
+use with many off-the-shelf algorithms avaialable in the RLLib
+framework.
+
+Before jumping into the implementation details, let’s take a look a
+simple motivating example: the hallway problem.
+
+The hallway problem is effectively a 1d version of the gridworld problem
+in classic RL. We are given a hallway with :math:`N` discrete positions
+and, starting at one end, want to learn to reach the opposite end in as
+few steps as possible.
+
+.. figure:: ./img/hallway-flat.png
+   :alt: hallway-flat
+
+   hallway-flat
+
+The figure above shows a hallway problem with :math:`N=3`, and the
+optimal solution starting at state 0 and ending at state 2, with each
+“current state” highlighted in black.
+
+This trivial problem can be used to express the “RL on a graph” idea
+succinctly, and enable solving much more interesting, non-trivial
+problems.
+
+The hallway problem as graph problem
+------------------------------------
+
+Before we jump into the graph formulation of the hallway problem, let’s
+talk first about *actions*, because this is one of the key differences
+between ``graphenv`` and traditional RL gym environments.
+
+Typically, discrete actions spaces in gym environments have fixed
+dimension and fixed definitions. In the hallway problem, for instance,
+there are always two actions: “move left” (action=0) and “move right”
+(action=1). Cases where the action is infeasible, like trying to move
+left from the start position, are typically handled by implmenting a
+*null* transition where the action doesn’t change the current state.
+
+In general graph search problems, however, such fixed action spaces are
+not practical. In the game of chess, for example, the total number of
+possible board states and, subsequently, moves, is astronomical, while
+the set of *feasible* moves changes continually throughout the game.
+
+The key observation that makes graph search tractable as a gym
+environment is: even for large and complex graph search problems, the
+number of states that are accessible from the current state is usually
+relatively small. If, instead of fixing the action to a pre-defined set
+of choices, we think of each action as representing an accessible next
+state, this endeavor becomes tractable.
+
+And so, we abandon the idea of “fixed” action spaces in favor of
+“parametric” action spaces. Here, **parametric** means that actions
+coincide with next states represented by feature vectors, rather than
+having a single, index-based interpretation. In other words, rather than
+actions as “move :math:`[ left, right ]`”, parametric actions can be
+thought of as “go to state :math:`[i, j]`” where the states :math:`i,j`
+each have a vector representation. Also unliked fixed spaces, the order
+doesn’t matter: we could equivalently say “go to state :math:`[j, i]`”.
```

Comment:
Maybe not here, but somewhere in the docs (and likely in the corresponding paper) we'll want a *why graphenv?* section. Rllib does have some examples with parametric actions and masked actions, but we're (hopefully) making those concepts much easier for a user. (The rllib parametric actions demo just embeds (left, right) with static feature vectors for a cart-pole, for instance)

Reply:
i agree!  the scope for this notebook was more about the simple example, but i agree we need a motivation section -- perhaps just one click from the landing page.  

yeah, that's a good distinction, and i was struggling with how to express this.  we are different in that 

* actions <-> next states, rather than fixed actions conditioned on the current state (as in cartpole)
* the implementation is much simpler -- just define an obs space for a vertex and a model that is aware of it, the rest is done for the user